File: Sources/Olaf/Core/Engine/OlafRuntime.swift
import Foundation

/// The state owner behind the `Olaf` facade. Guards the store's lifecycle, the kill switch, and
/// the level threshold with a lock. `@unchecked Sendable` — all mutable state sits behind `lock`.
final class OlafRuntime: @unchecked Sendable {

    private let lock = NSLock()
    private var _store: LogStore?
    private var _minimumLevel: LogLevel = .debug
    private var _enabled = true
    private var _sessionID = ""

    /// Logs emitted BEFORE `start()` is called are buffered here and flushed on start.
    /// (So early logs during app launch — e.g. splash — aren't lost.)
    private var _pending: [PendingLog] = []
    private let maxPending = 1000

    private struct PendingLog {
        let date: Date
        let level: LogLevel
        let category: LogCategory
        let rawMessage: String
        let rawMetadata: [String: String]
        let file: String
        let line: Int
        let function: String
        let thread: String
    }

    /// Where a log call should go.
    enum LogTarget {
        case store(LogStore)   // started + passed the level threshold → write directly
        case buffer            // not yet started → buffer (flushed on start)
        case drop              // disabled or below the level threshold → discard
    }

    // MARK: - Lifecycle

    /// Idempotent start. The first call wins.
    func start(with configuration: OlafConfiguration) {
        lock.lock(); defer { lock.unlock() }
        guard _store == nil else { return }

        _sessionID = Self.makeSessionID()

        let persistence: FilePersistence?
        if configuration.persistsToDisk {
            persistence = FilePersistence(
                directory: Self.defaultLogDirectory(),
                maxFileSize: configuration.maxFileSize,
                maxFileCount: configuration.maxFileCount
            )
        } else {
            persistence = nil
        }

        let mirror = configuration.mirrorsToOSLog
            ? OSLogMirror(subsystem: configuration.subsystem)
            : nil

        _store = LogStore(
            capacity: configuration.inMemoryCapacity,
            persistence: persistence,
            exportFormatter: configuration.exportFormatter,
            osLogMirror: mirror,
            sessionID: _sessionID
        )
        _minimumLevel = configuration.minimumLevel

        // Flush logs buffered before start (according to the level threshold).
        if let store = _store, !_pending.isEmpty {
            for pending in _pending where pending.level >= _minimumLevel {
                store.ingest(
                    date: pending.date,
                    level: pending.level,
                    category: pending.category,
                    rawMessage: pending.rawMessage,
                    rawMetadata: pending.rawMetadata,
                    file: pending.file,
                    line: pending.line,
                    function: pending.function,
                    thread: pending.thread
                )
            }
            _pending.removeAll()
        }
    }

    // MARK: - Access

    var store: LogStore? {
        lock.lock(); defer { lock.unlock() }
        return _store
    }

    var isStarted: Bool {
        lock.lock(); defer { lock.unlock() }
        return _store != nil
    }

    /// Current session identifier (populated after `start()`; empty before that).
    var currentSessionID: String {
        lock.lock(); defer { lock.unlock() }
        return _sessionID
    }

    var isEnabled: Bool {
        get { lock.lock(); defer { lock.unlock() }; return _enabled }
        set { lock.lock(); _enabled = newValue; lock.unlock() }
    }

    /// Collection threshold. Comes from the `start` config; can be changed at runtime
    /// (e.g. the "Collection threshold" setting in the viewer). Not persistent — scoped to the process lifetime.
    var minimumLevel: LogLevel {
        get { lock.lock(); defer { lock.unlock() }; return _minimumLevel }
        set { lock.lock(); _minimumLevel = newValue; lock.unlock() }
    }

    /// Determines the target of a log call (the message is only computed if not `.drop`).
    func target(for level: LogLevel) -> LogTarget {
        lock.lock(); defer { lock.unlock() }
        guard _enabled else { return .drop }
        if let store = _store {
            return level >= _minimumLevel ? .store(store) : .drop
        }
        return .buffer   // not yet started → buffer
    }

    /// Buffers a pre-start log (writes directly to the store if start was called in the meantime).
    func buffer(
        date: Date,
        level: LogLevel,
        category: LogCategory,
        rawMessage: String,
        rawMetadata: [String: String],
        file: String,
        line: Int,
        function: String,
        thread: String
    ) {
        lock.lock(); defer { lock.unlock() }
        if let store = _store {
            store.ingest(date: date, level: level, category: category, rawMessage: rawMessage, rawMetadata: rawMetadata, file: file, line: line, function: function, thread: thread)
            return
        }
        _pending.append(PendingLog(date: date, level: level, category: category, rawMessage: rawMessage, rawMetadata: rawMetadata, file: file, line: line, function: function, thread: thread))
        if _pending.count > maxPending {
            _pending.removeFirst(_pending.count - maxPending)
        }
    }

    // MARK: - Helpers

    /// Session identifier: a sortable timestamp-based prefix + a short random suffix.
    static func makeSessionID() -> String {
        "\(Int(Date().timeIntervalSince1970 * 1000))-\(UUID().uuidString.prefix(8))"
    }

    static func defaultLogDirectory() -> URL {
        let base = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        return base.appendingPathComponent("Olaf", isDirectory: true)
    }

    /// A readable label for the calling thread: "main" / thread name / dispatch queue label.
    static func currentThreadLabel() -> String {
        if Thread.isMainThread { return "main" }
        if let name = Thread.current.name, !name.isEmpty { return name }
        let label = String(cString: __dispatch_queue_get_label(nil))
        return label.isEmpty ? "background" : label
    }
}

File: Android/olaf/src/main/java/com/olaf/internal/OlafRuntime.kt
package com.olaf.internal

import android.content.Context
import com.olaf.LogCategory
import com.olaf.LogLevel
import com.olaf.OlafConfiguration
import java.io.File
import java.time.Instant
import java.util.UUID
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * The state owner behind the `Olaf` facade: guards the store's lifecycle, the kill switch and the
 * level threshold. All mutable state sits behind [lock].
 */
internal class OlafRuntime {

    private val lock = ReentrantLock()
    private var store: LogStore? = null
    private var minimum: LogLevel = LogLevel.DEBUG
    private var enabled = true
    private var sessionId = ""
    private var cacheDirectory: File? = null
    private var notifier: OlafNotifier? = null

    /**
     * Logs emitted BEFORE `start()` are buffered here and flushed on start, so early launch
     * logging (splash, DI setup) isn't lost.
     */
    private val pending = ArrayDeque<PendingLog>()

    private class PendingLog(
        val date: Instant,
        val level: LogLevel,
        val category: LogCategory,
        val message: String,
        val metadata: Map<String, String>,
        val file: String,
        val line: Int,
        val function: String,
        val thread: String
    )

    /** Where a log call should go. */
    internal sealed interface LogTarget {
        /** Started and above the threshold → write straight through. */
        class Store(val store: LogStore) : LogTarget

        /** Not started yet → buffer, flushed on start. */
        data object Buffer : LogTarget

        /** Disabled or below the threshold → discard without computing the message. */
        data object Drop : LogTarget
    }

    // MARK: - Lifecycle

    /** Idempotent start — the first call wins. */
    fun start(context: Context, configuration: OlafConfiguration) {
        val appContext = context.applicationContext
        // The notifier needs a Context, so it is built here and handed to the store; the
        // Context-free `start` below stays usable from tests.
        notifier = if (configuration.showsNotification) {
            runCatching { OlafNotifier(appContext) }.getOrNull()
        } else {
            null
        }
        start(appContext.cacheDir, configuration)
    }

    /**
     * The Context-free half of [start] — everything Olaf needs from the app is its cache
     * directory, which also keeps the engine unit-testable without an Android runtime.
     */
    fun start(cacheDir: File, configuration: OlafConfiguration) {
        lock.withLock {
            if (store != null) return

            sessionId = makeSessionId()
            cacheDirectory = cacheDir

            val persistence = if (configuration.persistsToDisk) {
                FilePersistence.create(
                    directory = File(cacheDir, LOG_DIRECTORY_NAME),
                    maxFileSize = configuration.effectiveMaxFileSize,
                    maxFileCount = configuration.effectiveMaxFileCount,
                    retentionMillis = configuration.retentionMillis
                )
            } else {
                null
            }

            val mirror = if (configuration.mirrorsToLogcat) LogcatMirror(configuration.logcatTag) else null

            val created = LogStore(
                capacity = configuration.inMemoryCapacity,
                persistence = persistence,
                exportFormatter = configuration.exportFormatter,
                logcatMirror = mirror,
                notifier = notifier,
                sessionId = sessionId,
                cacheDirectory = cacheDir
            )
            store = created
            minimum = configuration.minimumLevel

            // Flush what was buffered before start, honouring the threshold.
            while (pending.isNotEmpty()) {
                val log = pending.removeFirst()
                if (log.level >= minimum) {
                    created.ingest(
                        date = log.date,
                        level = log.level,
                        category = log.category,
                        message = log.message,
                        metadata = log.metadata,
                        file = log.file,
                        line = log.line,
                        function = log.function,
                        thread = log.thread
                    )
                }
            }
        }
    }

    // MARK: - Access

    fun store(): LogStore? = lock.withLock { store }

    fun cacheDirectory(): File? = lock.withLock { cacheDirectory }

    /** Clears the capture notification — the viewer calls this once it is on screen. */
    fun dismissNotification() {
        lock.withLock { notifier }?.dismiss()
    }

    val isStarted: Boolean get() = lock.withLock { store != null }

    /** Current session identifier — populated by [start], empty before that. */
    val currentSessionId: String get() = lock.withLock { sessionId }

    var isEnabled: Boolean
        get() = lock.withLock { enabled }
        set(value) = lock.withLock { enabled = value }

    /**
     * Collection threshold. Seeded from the start configuration and changeable at runtime (the
     * viewer exposes it). Not persisted — scoped to the process lifetime.
     */
    var minimumLevel: LogLevel
        get() = lock.withLock { minimum }
        set(value) = lock.withLock { minimum = value }

    /** Decides where a log call goes; the message is only computed when this isn't [LogTarget.Drop]. */
    fun target(level: LogLevel): LogTarget = lock.withLock {
        if (!enabled) return LogTarget.Drop
        val current = store ?: return LogTarget.Buffer
        if (level >= minimum) LogTarget.Store(current) else LogTarget.Drop
    }

    /** Buffers a pre-start log (writes straight through if start happened in the meantime). */
    fun buffer(
        date: Instant,
        level: LogLevel,
        category: LogCategory,
        message: String,
        metadata: Map<String, String>,
        file: String,
        line: Int,
        function: String,
        thread: String
    ) {
        lock.withLock {
            val current = store
            if (current != null) {
                current.ingest(date, level, category, message, metadata, file, line, function, thread)
                return
            }
            pending.addLast(
                PendingLog(date, level, category, message, metadata, file, line, function, thread)
            )
            while (pending.size > MAX_PENDING) {
                pending.removeFirst()
            }
        }
    }

    private companion object {
        const val MAX_PENDING = 1000
        const val LOG_DIRECTORY_NAME = "olaf"

        /** Sortable timestamp prefix plus a short random suffix. */
        fun makeSessionId(): String =
            "${System.currentTimeMillis()}-${UUID.randomUUID().toString().take(8)}"
    }
}
